Mathlib/MeasureTheory/SetAlgebra.lean

```diff
@@ -0,0 +1,227 @@
+theorem univ_mem (h𝒜 : IsSetAlgebra 𝒜) : univ ∈ 𝒜 :=
+  compl_empty ▸ h𝒜.compl_mem h𝒜.empty_mem
+
+/-- An algebra of sets is closed by intersection. -/
+theorem inter_mem (h𝒜 : IsSetAlgebra 𝒜) (s_mem : s ∈ 𝒜) (t_mem : t ∈ 𝒜) :
+    s ∩ t ∈ 𝒜 :=
+  inter_eq_compl_compl_union_compl .. ▸
+    h𝒜.compl_mem (h𝒜.union_mem (h𝒜.compl_mem s_mem) (h𝒜.compl_mem t_mem))
+
+/-- An algebra of sets is closed by difference. -/
+theorem diff_mem (h𝒜 : IsSetAlgebra 𝒜) (s_mem : s ∈ 𝒜) (t_mem : t ∈ 𝒜) :
+    s \ t ∈ 𝒜 := h𝒜.inter_mem s_mem (h𝒜.compl_mem t_mem)
+
+/-- An algebra of sets is a ring of sets. -/
+theorem isSetRing (h𝒜 : IsSetAlgebra 𝒜) : IsSetRing 𝒜 where
+  empty_mem := h𝒜.empty_mem
+  union_mem := h𝒜.union_mem
+  diff_mem := fun _ _ ↦ h𝒜.diff_mem
+
+/-- An algebra of sets is closed by finite unions. -/
+theorem biUnion_mem {ι : Type*} (h𝒜 : IsSetAlgebra 𝒜) {s : ι → Set α} (S : Finset ι)
+    (hs : ∀ i ∈ S, s i ∈ 𝒜) : ⋃ i ∈ S, s i ∈ 𝒜 := h𝒜.isSetRing.biUnion_mem S hs
+
+/-- An algebra of sets is closed by finite intersections. -/
+theorem biInter_mem {ι : Type*} (h𝒜 : IsSetAlgebra 𝒜) {s : ι → Set α} (S : Finset ι)
+    (hs : ∀ i ∈ S, s i ∈ 𝒜) : ⋂ i ∈ S, s i ∈ 𝒜 := by
+  by_cases h : S = ∅
+  · rw [h, ← Finset.set_biInter_coe, Finset.coe_empty, biInter_empty]
+    exact h𝒜.univ_mem
+  · rw [← ne_eq, ← Finset.nonempty_iff_ne_empty] at h
+    exact h𝒜.isSetRing.biInter_mem S h hs
+
+end IsSetAlgebra
+
+section generateSetAlgebra
+
+/-- `generateSetAlgebra 𝒜` is the smallest algebra of sets containing `𝒜`. -/
+inductive generateSetAlgebra {α : Type*} (𝒜 : Set (Set α)) : Set (Set α)
+  | base {s : Set α} (s_mem : s ∈ 𝒜) : generateSetAlgebra 𝒜 s
+  | empty : generateSetAlgebra 𝒜 ∅
+  | compl {s : Set α} (hs : generateSetAlgebra 𝒜 s) : generateSetAlgebra 𝒜 sᶜ
+  | union {s t : Set α} (hs : generateSetAlgebra 𝒜 s) (ht : generateSetAlgebra 𝒜 t) :
+      generateSetAlgebra 𝒜 (s ∪ t)
+
+/-- The algebra of sets generated by a family of sets is an algebra of sets. -/
+theorem isSetAlgebra_generateSetAlgebra :
+    IsSetAlgebra (generateSetAlgebra 𝒜) where
+  empty_mem := generateSetAlgebra.empty
+  compl_mem := fun _ ↦ generateSetAlgebra.compl
+  union_mem := fun _ _ ↦ generateSetAlgebra.union
+
+/-- The algebra of sets generated by `𝒜` contains `𝒜`. -/
+theorem self_subset_generateSetAlgebra : 𝒜 ⊆ generateSetAlgebra 𝒜 :=
+  fun _ ↦ generateSetAlgebra.base
+
+/-- The measurable space generated by a family of sets `𝒜` is the same as the one generated
+by the algebra of sets generated by `𝒜`. -/
+theorem generateFrom_generateSetAlgebra_eq :
+    generateFrom 𝒜 = generateFrom (generateSetAlgebra 𝒜) := by
+  refine le_antisymm (generateFrom_mono self_subset_generateSetAlgebra) (fun s ms ↦ ?_)
+  refine @generateFrom_induction _ _ (generateSetAlgebra 𝒜) (fun t ht ↦ ?_)
+    (@MeasurableSet.empty _ (generateFrom 𝒜))
+    (fun t ↦ MeasurableSet.compl)
+    (fun f hf ↦ MeasurableSet.iUnion hf)
+    s ms
+  induction' ht with u u_mem u _ mu u v _ _ mu mv
```

Comment:
Could you switch this to the unprimed `induction` tactic, which is a bit more readable?